### PR TITLE
MINOR: Save failed test output to build output directory

### PR DESCRIPTION
**Pull request by ewencp**
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by ewencp** [comment]
We should be able to also update the Jenkins jobs to collect the files, just save anything matching `**/testOutput/` or with the extra `*.stdout` if we want to be more targeted (I'm not sure if anything else uses that subdirectory).

**Comment by ewencp** [comment]
retest this please

**Comment by ewencp** [comment]
ping. it'd be good to get this in since it'll help with debugging tests that flake pretty consistently on jenkins (though we may also need some follow up to turn up logging output during tests, I think a bunch of modules don't even have logging turned on).

File: build.gradle [hunk 1]
```diff
@@ -217,6 +223,28 @@ subprojects {
       exceptionFormat = testExceptionFormat
     }
 
+    // Gradle built-in logging only supports sending test output to stdout, which generates a lot
+    // of noise, especially for passing tests. We really only want output for failed tests. This
+    // hooks into the output and logs it (so we don't have to buffer it all in memory) and only
+    // saves the output for failing tests. Directory and filenames are such that you can, e.g.,
+    // create a Jenkins rule to collect failed test output.
+    def logFiles = new HashMap<TestDescriptor, File>()
+    def logStreams = new HashMap<TestDescriptor, FileOutputStream>()
```
Comment (cmccabe) [inline build.gradle]: Should this be in a try/catch block to avoid leaks?

File: build.gradle [hunk 2]
```diff
@@ -217,6 +223,28 @@ subprojects {
...
+      logStreams.put(testId(td), new FileOutputStream(logFile))
+    }
+    onOutput { TestDescriptor td, TestOutputEvent toe ->
+      logStreams.get(testId(td)).write(toe.getMessage().getBytes(StandardCharsets.UTF_8))
+    }
+    afterTest { TestDescriptor td, TestResult tr ->
+      logStreams.get(testId(td)).close()
+      if (tr.resultType != TestResult.ResultType.FAILURE) {
+        logFiles.get(testId(td)).delete()
+      }
```
Comment (cmccabe) [inline build.gradle]: Should we print out the name of the file if there is a failure?  It might be helpful for people invoking this from the command line.

**Comment by cmccabe** [comment]
Thanks, @ewencp, this will be extremely helpful.  LGTM after comments are addressed.

**Comment by ewencp** [comment]
@cmccabe thanks for reviewing. I had to rework a bit because I realized we need to repeat everything in each of the `test`, `unitTest`, and `integrationTest` stanzas. I'm no groovy expert, so not really sure I did this in the best way.

Also added some error handling, and some edge cases that seem to be due to Gradle brokenness, so left long comment explaining (hopefully clearly enough) what's going on.

**Comment by ewencp** [comment]
Given previous +1, I'm going to merge this to trunk so we can start getting these files to help with debugging. Assuming everything plays out well, I will later backport as far as reasonable (hopefully pretty much all active branches back to 1.0, depending on any cherry-pick complications).

**Pull request closed without merging**